README.md
# Bhulk - Bulk requests for hapi

Bulk request plugin for hapi. Bhulk is an alternative to [bassmaster](https://github.com/spumko/bassmaster) and the main difference is that Bhulk only performs and responds to GET-requests. So in situations where you want a way to fetch things in bulk in a more RESTful way Bhulk might be a good alternative.

# Using the plugin

Install bhulk in your project `npm install -S bhulk` and then require it into your server:

```javascript
server.pack.require('bhulk', {
  path: '/bulk',
}, function setupDone(error) {
  if (error) throw error;
  console.log('Bhulk smash!');
})
```

Passing the path:'/bulk' option is not necessary as it is the default. But there are a couple of options for controlling the behaviour of Bhulk:

  * `path` - the path for the Bhulk route, defaults to '/bulk'.
  * `routeConfig` - configuration options for the route, will be merged with the default route config, see the "config" section of [Route options](https://github.com/spumko/hapi/blob/master/docs/Reference.md#route-options) in the hapi docs.
  * `remote` - by default Bhulk only performs subrequests to the server it's configured for, but by passing in a `remote` you can specify a function that can call remote services as part of a bulk request. The signature of the function is `remote(remote, request, reply)`. The request object has two attributes: `url` and `credentials`.
  * `iterationLimit` - the maximum number of iteration requests that should be allowed, defaults to `10`.

# Performing bulk requests

A sample use of Bhulk (the query parameters as json for easier reading):

```json
{
  "stories": "/collections/5/stories?fields=entityId,entityType,addedBy",
  "posts": "/posts?ids=${,}&fields=title,url",
  "posts.source": "stories",
  "posts.query": ".[*](eq(@.entityType, 'post')).entityId",
  "authors": "/users?ids=${,}",
  "authors.source": "stories",
  "authors.query": ".*.addedBy",
  "_debug": "true"
}
```

All requests in Bhulk are named here we have "stories", "posts", and "authors". A request can have a `source`, which is a dependency on another request, whose result then can be used to build the request. The what values to select from the sources result is determined by the `query`, which is an [OBPath](https://www.npmjs.org/package/obpath.js) expression. So in the above exaple the "posts" request has "stories" and a `source` and it will take all "entityId" values from objects in the result that have an "entityType" that matches 'post'. The `${,}` placeholder in the "posts" url "/posts?ids=${,}&fields=title,url,content,image" will be replaced with the unique values from the JSONPath query, separated by comma. So the actual request that will be made could look something like this: "/posts?ids=123,234&fields=title,url,content,image". It's also possible to suppress the results of requests with a suppress parameter, for stories that would look like this: "stories.suppress=true".

The "_debug" parameter causes bhulk to output metadata about the requests; how they are interpreted and what the final url resolves to.

Requests will be executed in parallel when possible. The "posts" and "authors" requests in the above example will both be made as soon as their `source` request has finished.

The result will look like this:

```json
{
  "meta": {
    "requests": {
      "stories": {
        "url": "/collections/5/stories?fields=entityId,entityType,addedBy"
      },
      "posts": {
        "url": "/posts?ids=${,}&fields=title,url",
        "source": "stories",
        "query": ".[*](eq(@.entityType, 'post')).entityId",
        "realUrl": "/posts?ids=123,234&fields=title,url"
      },
      "authors": {
        "url": "/users?ids=${,}",
        "source": "stories",
        "query": ".*.addedBy",
        "realUrl": "/users?ids=1001"
      }
    }
  },
  "results": {
    "stories": [
      {"entityId": 123, "entityType": "post", "addedBy": 1001},
      {"entityId": 234, "entityType": "post", "addedBy": 1001}
    ],
    "posts": [
      {"title":"Foo", "url":"http://example.com/foo"},
      {"title":"Bar", "url":"http://example.com/bar"}
    ],
    "authors": [
      {"id": 1001, "name": "Jane Doe", "avatar": "http://example.com/jane.png"}
    ]
  }
}
```

## Iteration

In cases where it isn't practical or possible to make a bulk request, like the one in the posts example above, it's also possible to make one separate request per path expression match. This should be used sparingly, bulk loading is always preferred.

A sample use of iteration (the query parameters as json for easier reading):

```json
{
  "collections": "/users/74841/collections?count=10",
  "collections.remote": "v1-collections",
  "stories": "/collections/${.}/stories?count=3",
  "stories.remote": "v1-collections",
  "stories.each": "true",
  "stories.source": "collections",
  "stories.query": ".result[*].id",
  "posts": "/posts?ids=${,}&fields=id,url,title,image",
  "posts.remote": "v1-core",
  "posts.source": "stories",
  "posts.query": "[*].result.result[*].storyId"
}
```

Here we fetch a users ten first collections (remember the `iterationLimit` here). Then we issue a request for the three first stories for each of the returned collections. As this isn't a "bulk" request we're not joining the result of the obpath query, The "${.}" (period instead of comma) placeholder uses the matched item directly instead.

The posts request then collects all the storyIds, and loads the corresponding posts with a bulk load requests.

```json
{
  "meta": {
    "requests": {
      "collections": {
        "remote": "v1-collections",
        "url": "/users/74841/collections?count=10"
      },
      "stories": {
        "each": true,
        "query": ".result[*].id",
        "url": "/collections/${.}/stories?count=3",
        "remote": "v1-collections",
        "realUrls": [
          "/collections/53f60850458a0c81cbe8d7b2/stories?count=3",
          "/collections/53f5f9d6c651260ebf21c255/stories?count=3",
          "/collections/53ec78f8bfb0999d51ec95a0/stories?count=3"
        ],
        "source": "collections"
      },
      "posts": {
        "url": "/posts?ids=${,}&fields=id,url,title,image",
        "remote": "v1-core",
        "source": "stories",
        "realUrl": "/posts?ids=2576%2C123&fields=id,url,title,image",
        "query": "[*].result.result[*].storyId"
      }
    }
  },
  "results": {
    "collections": {
      "meta": {},
      "result": [
        {
          "authors": [
            {
              "added": 1408632912901,
              "level": 1,
              "id": "53f60850458a0c81cbe8d7b3",
              "userId": 74841
            }
          ],
          "created": 1408632912901,
          "id": "53f60850458a0c81cbe8d7b2",
          "followers": [],
          "name": "Collection with posts",
          "public": true,
          "modified": 1408632912901,
          "version": 0
        },
        {
          "version": 0,
          "public": true,
          "modified": 1408629206001,
          "name": "Collection with posts",
          "id": "53f5f9d6c651260ebf21c255",
          "followers": [
            "12788",
            "74841"
          ],
          "created": 1408629206001,
          "authors": [
            {
              "userId": 74841,
              "added": 1408629206001,
              "level": 1,
              "id": "53f5f9d6c651260ebf21c256"
            }
          ]
        },
        {
          "version": 0,
          "public": true,
          "modified": 1408006392775,
          "name": "Awesome collection",
          "id": "53ec78f8bfb0999d51ec95a0",
          "followers": [
            "12788",
            "74841"
          ],
          "created": 1408006392775,
          "authors": [
            {
              "userId": 74841,
              "added": 1408006392775,
              "id": "53ec78f8bfb0999d51ec95a1",
              "level": 1
            }
          ]
        }
      ]
    },
    "stories": [
      {
        "result": {
          "meta": {},
          "result": [
            {
              "addedBy": 74841,
              "public": true,
              "collectionId": "53f60850458a0c81cbe8d7b2",
              "version": 0,
              "storyId": "2576",
              "id": "53f6086f458a0c81cbe8d7b4",
              "added": 1408632943924
            }
          ]
        },
        "from": "53f60850458a0c81cbe8d7b2"
      },
      {
        "from": "53f5f9d6c651260ebf21c255",
        "result": {
          "result": [
            {
              "added": 1408629219268,
              "id": "53f5f9e3c651260ebf21c257",
              "version": 0,
              "storyId": "2576",
              "collectionId": "53f5f9d6c651260ebf21c255",
              "public": true,
              "addedBy": 74841
            }
          ],
          "meta": {}
        }
      },
      {
        "result": {
          "meta": {
            "nextUrl": "/collections/53ec78f8bfb0999d51ec95a0/stories?count=3&before=53f5e8366ceaa8c9a5cc2eca"
          },
          "result": [
            {
              "addedBy": 74841,
              "public": true,
              "version": 0,
              "collectionId": "53ec78f8bfb0999d51ec95a0",
              "storyId": "123",
              "id": "542bf8f63bc045cd4e210336",
              "added": 1412167926784
            },
            {
              "public": true,
              "addedBy": 74841,
              "added": 1412167854006,
              "version": 0,
              "collectionId": "53ec78f8bfb0999d51ec95a0",
              "storyId": "2576",
              "id": "542bf8ae3bc045cd4e210335"
            },
            {
              "added": 1408624694894,
              "id": "53f5e8366ceaa8c9a5cc2eca",
              "collectionId": "53ec78f8bfb0999d51ec95a0",
              "version": 0,
              "storyId": "2576",
              "public": true,
              "addedBy": 74841
            }
          ]
        },
        "from": "53ec78f8bfb0999d51ec95a0"
      }
    ],
    "posts": [
      {
        "id": "2576",
        "title": "11e juli 2012 – på Sardinien",
        "url": "http://kenzas.se/2012/07/11/11e-juli-2012-pa-sardinien/",
        "image": "http://kenzas.se/wp-content/uploads/2012/07/IMG_1672.jpg"
      },
      {
        "id": "123",
        "url": "http://emmatokfrans.blogg.se/2011/august/vardag.html",
        "image": "http://emmatokfrans.blogg.se/images/2011/dsc01249_162112190.jpg",
        "title": "Vardag"
      }
    ]
  }
}
```


CHANGELOG.md
## 1.2.0 (2015-01-19)


#### Features

* **main:** detect errors and empty results ([c376b58e](https://github.com/bloglovin/bhulk/commit/c376b58e1f935f9ebfc8d71b23c6cac396d05730))


### 1.1.3 (2014-12-03)


#### Bug Fixes

* **main:** set a request name ([e53bb981](https://github.com/bloglovin/bhulk/commit/e53bb9818bb24ba5dc6a0d17f90dbfd09d06a31a))


### 1.1.2 (2014-10-22)


#### Bug Fixes

* **obpath:** updated obpath.js to 1.x ([95967f6e](https://github.com/bloglovin/bhulk/commit/95967f6e4a860b1951a7e7f7b48aa344247ebea0))


### 1.1.1 (2014-10-21)


#### Bug Fixes

* **plugin:** properly register hapi plugin metadata ([b2707074](https://github.com/bloglovin/bhulk/commit/b270707443ae8fad4dd46006caddb9705c48a668))


## 1.1.0 (2014-10-10)


#### Features

* **iteration:** allow forEach-style requests ([7ba5b11d](https://github.com/bloglovin/bhulk/commit/7ba5b11d44c6b679d270614d47ccb1e5094758fd))



index.js
/* jshint node: true */
'use strict';

var mod_async = require('async');
var mod_obpath = require('obpath.js');

var _ = require('lodash');

function Bhulk(hapi, config) {
  this.hapi = hapi;
  this.config = config;
  this.obpathContext = mod_obpath.createContext();
  this.queryCache = {};
}

Bhulk.prototype.bulkRequest = function (bulkRequest, reply) {
  var bulkResult = {
    meta: {}
  };
  var hapi = this.hapi;
  var requests = {};
  var remoteHandler = this.config.remote;
  var iterationLimit = this.config.iterationLimit || 10;

  for (var key in bulkRequest.query) {
    if (key.substr(0,1) === '_') { continue; }

    var segments = key.split('.');
    var value = bulkRequest.query[key];
    var request = requests[segments[0]] = requests[segments[0]] || {};

    if (segments.length === 1) {
      request.name = segments[0];
      request.url = value;
    }
    else if (segments.length === 2) {
      if (segments[1] === 'source') {
        request.source = value;
      }
      else if (segments[1] === 'each') {
        if (value === 'true') {
          request.each = true;
        }
      }
      else if (segments[1] === 'suppress') {
        if (value === 'true') {
          request.suppress = true;
        }
      }
      else if (segments[1] === 'query') {
        if (!this.queryCache[value]) {
          try {
            this.queryCache[value] = mod_obpath.mustCompile(value, this.obpathContext);
          } catch (syntaxError) {
            reply(hapi.error.badRequest('Bad path expression in "' + key + '": ' + syntaxError.message));
            return;
          }
        }
        request.query = this.queryCache[value];
      }
      else if (segments[1] === 'remote') {
        request.remote = value;
      }
    }
  }

  var treeError = this.validateTree(requests);
  if (treeError) {
    reply(hapi.error.badRequest(treeError.message));
    return;
  }

  var tasks = {};

  var urlReplacement = /\$\{([^}])\}/;
  function replace(placeholder, params) {
    var value = '';
    if (placeholder === ',') {
      value = _.uniq(params).join(placeholder);
    }
    if (placeholder === '.') {
      value = params.toString();
    }
    return encodeURIComponent(value.toString());
  }

  function requestWorker(requestSpec, callback, sources) {
    var data, params;

    if (requestSpec.source && requestSpec.query) {
      data = sources[requestSpec.source];

      // Check if the source request was successful
      if (requestSpec.source && (!data || (data.error && data.statusCode))) {
        if (data.statusCode === 412) {
          return callback(null, {
            error: 'Dependency Failure',
            statusCode: 412,
            message: 'Skipping request as the source request was skipped',
          });
        } else if (data.statusCode >= 400) {
          return callback(null, {
            error: 'Dependency Failure',
            statusCode: 412,
            message: 'Skipping request as the source request failed',
          });
        }
      }

      params = requestSpec.query.evaluate(data);
      if (!params || params.length === 0) {
        return callback(null, {
          error: 'Dependency Failure',
          statusCode: 412,
          message: 'Skipping request as the source query didn\'t return any results',
        });
      }
    }

    if (!requestSpec.each) {
      var url = requestSpec.url;

      if (params) {
        url = requestSpec.url.replace(urlReplacement, function(match, placeholder) {
          return replace(placeholder, params);
        });
        if (url !== requestSpec.url) {
          requestSpec.realUrl = url;
        }
      }

      performRequest(requestSpec, url, callback);
    }
    else {
      if (params) {
        requestSpec.realUrls = [];

        if (params.length === 0) {
          return callback(null, {});
        }
        // We limit the number of request iterations we allow.
        else if (params.length > iterationLimit) {
          requestSpec.skippedUrls = params.slice(iterationLimit).map(function mapUrl(item) {
            return requestSpec.url.replace(urlReplacement, function(match, placeholder) {
              return replace(placeholder, item);
            });
          });
          params = params.slice(0, iterationLimit);
        }

        var queue = mod_async.queue(function eachJob(item, callback) {
          var url = requestSpec.url.replace(urlReplacement, function(match, placeholder) {
            return replace(placeholder, item);
          });
          requestSpec.realUrls.push(url);

          performRequest(requestSpec, url, function resultPair(err, result) {
            var pair = {
              from: item,
            };
            if (err) {
              pair.error = err;
            }
            else if (result) {
              pair.result = result;
            }
            callback(null, pair);
          });
        }, 3);

        var results = [];
        queue.push(params, function (err, result) {
          results.push(result);
        });

        queue.drain = function queueDone() {
          callback(null, results);
        };
      }
      else {
        callback(new Error('Iteration requests must have a source and query'));
      }
    }
  }

  function performRequest(requestSpec, url, callback) {
    if (requestSpec.remote) {
      if (typeof remoteHandler !== 'function') {
        callback(new Error('This endpoint doesn\'t support remote requests'));
        return;
      }
      remoteHandler(requestSpec.remote, {
        name: requestSpec.name,
        url: url,
        credentials: bulkRequest.auth.credentials
      }, function handleReply(reply) {
        if (reply && reply.isBoom) {
          reply = reply.output.payload;
        }
        callback(null, reply);
      });
    }
    else {
      var injectOptions = {
        url: url,
        method: 'GET',
        credentials: bulkRequest.auth.credentials
      };

      bulkRequest.server.inject(injectOptions, function(data) {
        callback(null, data.result);
      });
    }
  }

  // Build our task object
  for (var resourceName in requests) {
    var requestSpec = requests[resourceName];
    var worker = requestWorker.bind(this, requestSpec);

    var item;
    if (requestSpec.source) {
      item = [requestSpec.source, worker];
    }
    else {
      item = worker;
    }
    tasks[resourceName] = item;
  }


  mod_async.auto(tasks, function(error, results) {
    if (error) {
      reply(hapi.error.badRequest(error.message));
      return;
    }

    // Add debugging info
    if (bulkRequest.query._debug === 'true') {
      // Restore the path expressions
      for (var key in requests) {
        if (requests[key].query) {
          requests[key].query = requests[key].query.path;
        }
      }

      bulkResult.meta.requests = requests;
    }

    bulkResult.results = {};

    // Add the non-suppressed results
    for (var rkey in results) {
      if (!requests[rkey].suppress) {
        bulkResult.results[rkey] = results[rkey];
      }
    }

    reply(bulkResult);
  });
};

Bhulk.prototype.validateTree = function (requests) {
  var walked, current;
  for (var key in requests) {
    walked = [key];
    current = requests[key];
    while (current.source) {
      if (!requests[current.source]) {
        return new Error('Undefined request source "' + current.source + '"');
      }
      if (walked.indexOf(current.source) > -1) {
        walked.push(current.source);
        return new Error('Dependency loop: ' + walked.join(' -> '));
      }
      walked.push(current.source);
      current = requests[current.source];
    }
  }

  return null;
};

exports.register = function (plugin, options, next) {
  var config = {
    path: '/bulk',
    routeConfig: {
      description: 'Bulk request',
      notes: 'Performs multiple requests'
    },
    remote: null
  };
  if (options) {
    _.merge(config, options);
  }

  var bulk = new Bhulk(plugin.hapi, config);
  plugin.route({
    method: 'GET',
    path: config.path,
    handler: bulk.bulkRequest.bind(bulk),
    config: config.routeConfig
  });

  next();
};

exports.register.attributes = {
  pkg: require('./package'),
};


examples/bulk-with-remote.js
/* jshint node: true */
'use strict';

var lib = {
  async: require('async'),
  hapi: require('hapi'),
  http: require('http'),
  url: require('url'),
  request: require('request'),
  bhulk: require('../')
};

var serverOne = lib.hapi.createServer('0.0.0.0', 20180);
var serverTwo = lib.hapi.createServer('0.0.0.0', 20280);

serverOne.route([
  {
    path: '/listing',
    method: 'GET',
    handler: getListing
  },
  {
    path: '/entities',
    method: 'GET',
    handler: getEntities
  }
]);

serverTwo.route({
  path: '/users',
  method: 'GET',
  handler: getUsers
});

lib.async.auto({
  // Register the bhulk plugin
  bhulk: function registerBulk(callback) {
    // We have to register manually instead of using require as we're running from
    // inside bhulk.
    serverOne.pack.register({
      name: 'bhulk',
      plugin: lib.bhulk,
      options: {
        remote: handleRemoteRequest
      }
    }, callback);
  },
  // Start server one
  one: ['bhulk', function startOne(callback) {
    serverOne.start(callback);
  }],
  // Start server two
  two: function startTwo(callback) {
    serverTwo.start(callback);
  },
  // Make the sample request
  request: ['one', 'two', makeRequest]
}, function startResult(error, results) {
  if (error) {
    console.error('Failed to run example:', error);
    process.exit(1);
  }

  // Output the result if we were successful
  console.log(JSON.stringify(results.request, null, '  '));

  // Stop our servers
  serverOne.stop();
  serverTwo.stop();
});

// Makes the bulk request
function makeRequest(callback) {
  lib.request({
    url: 'http://127.0.0.1:20180/bulk',
    json: true,
    qs: {
      'listing': '/listing',
      'entities': '/entities?ids=${,}',
      'entities.source': 'listing',
      'entities.query': '.*.id',
      'authors': '/users?ids=${,}',
      'authors.remote': 'two',
      'authors.source': 'entities',
      'authors.query': '.*.by'
    }
  }, function requestResult(error, request, result) {
    callback(error, result);
  });
}

// Remote handler for server one, it only accepts one remote: 'two'.
function handleRemoteRequest(remote, request, reply) {
  if (remote !== 'two') {
    reply(lib.hapi.error.badRequest('Unknown remote "' + remote + '"'));
    return;
  }

  lib.request({
    url: lib.url.resolve('http://127.0.0.1:20280/', request.url),
    json: true
  }, function requestResult(error, request, result) {
    reply(result);
  });
}

/*
 * Mock route handlers that output some static data.
**/

function getListing(request, reply) {
  reply([
    {"id":123},
    {"id":234}
  ]);
}

function getEntities(request, reply) {
  var result = [];
  var db = {
    "123": {
      "id": 123,
      "name": "Foo",
      "by": 1001
    },
    "234": {
      "id": 234,
      "name": "Bar",
      "by": 1002
    }
  };

  var ids = request.query.ids.split(',');
  ids.forEach(function addResult(id) {
    if (db[id]) result.push(db[id]);
  });

  reply(result);
}

function getUsers(request, reply) {
  var result = [];
  var db = {
    "1001": {
      "id": 1001,
      "name": "Jane"
    },
    "1002": {
      "id": 1002,
      "name": "John"
    }
  };

  var ids = request.query.ids.split(',');
  ids.forEach(function addResult(id) {
    if (db[id]) result.push(db[id]);
  });

  reply(result);
}
